### ISSUE TITLE
Safely parsing samples out of `multiqc_data.json`

### ISSUE THREAD

**Issue by TMiguelT** [ewels/MegaQC#142]
Currently we have a problem with parsing samples out of `multiqc_data.json`. Currently we are taking every `name` field on any multiqc plot:
https://github.com/ewels/MegaQC/blob/d786c0fbe5e4a07098999ced99b32c76ff2465cc/megaqc/api/utils.py#L290-L292

However, this is a problem, because a single "sample" can end up having multiple `name` fields associated with it. For example, in the below snippet, we ran `fastqc` on a fastq, and then `multiqc`. 

```json
        "fastqc_adapter_content_plot": {
            "plot_type": "xy_line",
            "datasets": [
                [
                    {
                        "name": "PYV9385 - illumina_universal_adapter",
                        "data": [
                            [
                                1,
                                2.4949026222001998e-05
                            ],
                        ]
                    }
                ]
            ]
        }
```

The resulting name in the plot is `GIB18398 - illumina_universal_adapter`, which results in a new sample being imported, called `PYV9385 - illumina_universal_adapter`, which is not associated with the actual sample, called `PYV9385`. In addition, this `illumina_universal_adapter` sample appears in all the sample filter boxes etc, even though it has no other data associated it than a single plot.

Originally I thought this was a problem at the MultiQC end, but I now realise that this is simply the plotly JSON input, which needs to support having multiple adapter sequences per sample.

The question, then, is how do we handle this? We need this plot data to be associated with the correct sample, and to not create a dummy sample. This has implications on the MultiQC parser, and the data model.

I think we can reasonably model each of these plot datasets as an instance of `PlotData` (like it is currently), associated with the sample `PYV9385`, but we need to then add a metadata field that tells us which adapter it came from. In addition, we should probably check if these plot names are prefixed by an existing sample, and use that sample if so. This is a bit of a hack though, and I'd prefer a cleaner solution.

In general, we might need to add a module system like MultiQC to handle module-specific issues, because I can envisage naming behaviour being inconsistent across modules. But this would only be a last resort imo.

@ewels can you comment on this from a MultiQC perspective? I'm happy to do the work to fix this up, but I think I need some guidance here.

**Comment by ewels** [comment]
> Originally I thought this was a problem at the MultiQC end, but I now realise that this is simply the plotly JSON input, which needs to support having multiple adapter sequences per sample.

I have a feeling that this is the MultiQC end - plots have sample IDs but in this case there needs to be multiple plot lines for each sample. So the code makes multiple "samples" for each adapter so that the plot works. In an ideal world, the data structure would be able to handle this nesting whilst maintaining the main sample identifier. But I wrote this many years ago before MegaQC was even a twinkle in anyone's eye..

Note that I'm pretty sure that other modules do similar things, so it's probably not specific to just this plot. To make matters worse, I imagine that the naming conventions may not be the same everywhere.

**Comment by ewels** [comment]
I think that my take on it would be that it's not enough trouble to be worth it. MegaQC plots would hopefully still work fairly well with filters that match partial sample names. If you want a single statistic for monitoring etc then might be best to go to MultiQC to add that calculated average (eg. max adapter contamination, or whatever), so that it's reported with a single sample ID.

I think having a global fix for this will be messy and fragile, and maybe not worth the effort.

**Comment by TMiguelT** [comment]
Okay, so ideally the MultiQC data model would handle in a more structured way. I guess I can hope that will improve eventually, but maybe that's a big ask.

The problem here isn't really that we can't just use fuzzy filters. The problem is more that it's breaking the data model by introducing false samples.Your example about statistical monitoring is closer to our use-case, but in order for that to work, we need to be able to connect the `illumina_universal_adapter` with the underlying sample. With your knowledge of the modules and their idiosyncrasies, can you think of a more consistent way to do this? For example, are all the samples listed in one canonical place, and then we can assume that any plot names we find later on can be associated with an existing sample which is a substring of the plot name?

**Comment by ewels** [comment]
Sadly not - sample names are not and have never been canonical. This has always been a big stumbling block for MultiQC: figuring out what the sample name should be is super difficult. A fairly big chunk of MultiQC code/docs/my time spent is dedicated to sample name cleaning: https://multiqc.info/docs/#sample-name-cleaning

For example, FastQC produces reports per paired-end read so has two samples for each biological sample - how should that be handled? What prefixes can be removed, when is a sample two samples or one? The answers to each of these questions vary from person to person and use case to use case. I'm not sure that there is any right or wrong way to handle this kind of thing, without having some kind of sample sheet setup where you specify which biological sample every log file corresponds to. And that's not something that is ever going to happen.

**Comment by ewels** [comment]
So yeah, my approach has been to fudge it as best as I can with sensible defaults. But at it's heart, MultiQC sample names are not global across modules and not canonical in any way. Sorry 😉

**Comment by TMiguelT** [comment]
Right, so there's no canonical list of samples, but the sample cleaning mechanism shows that you do have some kind of internal rules for associating identifiers with samples. Since you have this list of samples in MultiQC, would it not be possible to fix this and other modules to app metadata to the plots to also include a canonical sample ID? [Plotly certainly supports metadata fields](https://plotly.com/python/reference/scatter/#scatter-meta). Such that my example becomes:

```json
        "fastqc_adapter_content_plot": {
            "plot_type": "xy_line",
            "datasets": [
                [
                    {
                        "name": "PYV9385 - illumina_universal_adapter",
                        "meta": {"sample_name": "PYV9385"},
                        "data": [
                            [
                                1,
                                2.4949026222001998e-05
                            ],
                        ]
                    }
                ]
            ]
        }
```
(not 100% sure this is the right JSON format, but it's the right idea)

**Comment by TMiguelT** [comment]
@ewels?

**Comment by ewels** [comment]
I guess this would be possible, but in 99% of cases it would be the same as the sample name. The case where MultiQC is _adding_ additional strings is very rare. This one FastQC plot does it, and I think that there are maybe one or two other plots across the hundreds that MultiQC generates.

Even if we added this, my point about the cleaning being far from perfect and not generating canonical sample names still stands. This metadata field will still contain the best-effort sample name which often has a lot of additional stuff in it and is not shared between modules.
